cron/src/main/scala/com/criteo/cuttle/cron/CronScheduler.scala

```diff
@@ -0,0 +1,155 @@
+  * Each [[com.criteo.cuttle.Job Job]] has it's own expression and executed separately from others.
+  *
+  * The [[com.criteo.cuttle.Scheduler Scheduler]] ensures that at least one
+  * [[com.criteo.cuttle.Execution Execution]] is created and successfully run for a given time instant.
+  * It also handles the retry policy.
+  *
+  * We follow the semantic of Cron on Unix systems and we don't manage misfires
+  * (meaning that if the scheduler missed some events because it was offline,
+  * or because a previous job execution was still running, we won't replay them).
+  *
+  * [[com.criteo.cuttle.Job Job]] is considered as finished when
+  * we can't produce the next computing instant from Cron expression.
+  */
+case class CronScheduler(logger: Logger) extends Scheduler[CronScheduling] {
+  private type CronExecution = Execution[CronScheduling]
+
+  private sealed trait CronJob {
+    val underlying: Job[CronScheduling]
+    def nextEvent = underlying.scheduling.nextEvent()
+  }
+  private case class New(override val underlying: Job[CronScheduling]) extends CronJob
+  private case class Retry(override val underlying: Job[CronScheduling], retryNum: Int, scheduledAt: ScheduledAt)
+      extends CronJob
+
+  override val name = "cron"
+
+  private val _state = Ref(Map.empty[CronJob, Either[Instant, CronExecution]])
+
+  private def addNextEventToState(job: CronJob, instant: Instant) = IO {
+    atomic { implicit txn =>
+      _state() = _state() + (job -> Either.left(instant))
+    }
+  }
+
+  private def addNextExecutionToState(job: CronJob, execution: CronExecution) = IO {
+    atomic { implicit txn =>
+      _state() = _state() + (job -> Either.right(execution))
+    }
+  }
+
+  private def removeJobFromState(job: CronJob) = IO {
+    atomic { implicit txn =>
+      _state() = _state() - job
+    }
+  }
+
+  private def logState() = IO {
+    val state = _state.single.get
+    logger.debug("======State======")
+    state.foreach {
+      case (job, jobState) =>
+        val messages = Seq(
+          job.underlying.id,
+          jobState.fold(_ toString, e => e.id)
+        )
+        logger.debug(messages mkString " :: ")
+    }
+    logger.debug("======End State======")
+  }
+
+  override def start(workload: Workload[CronScheduling],
+                     executor: Executor[CronScheduling],
+                     xa: XA,
+                     logger: Logger = logger): Unit = {
+    def runScheduledJob(job: CronJob, retryNum: Int, scheduledAt: ScheduledAt) =
+      for {
+        runInfo <- IO {
+          logger.debug(s"Sending job ${job.underlying.id} to executor")
+          val cronContext = CronContext(scheduledAt.instant)(retryNum)
+          executor.run(job.underlying, cronContext)
+        }
+        _ <- addNextExecutionToState(job, runInfo._1)
+        _ <- logState()
+        result <- IO.fromFuture(IO(runInfo._2))
+      } yield result
+
+    def run(job: CronJob): IO[Completed] = {
```

Comment:
I think there is a lot of duplication between `runScheduledJob` and `run`. I would refactor your code to keep a single more linear loop here. Something like:

```suggestion
def run(job: CronJob): IO[Unit] =
  for {
    maybeScheduledAt <- IO.pure(job.nextEvent)
    _ <-
      maybeScheduledAt match {
        // No next event. Stop the loop for this job
        case None =>
          logger.info(s"Job ${job.underlying.id} has finished. We will not submit executions anymore")
          IO.unit

        case Some(scheduledAt) =>

          // Launch an execution until success or retryNum is greater than maxRetry
          def runAndRetry(retryNum: Int): IO[Completed] =
            (for {
              runInfo <- IO {
                logger.debug(s"Sending job ${job.underlying.id} to executor")
                val cronContext = CronContext(scheduledAt.instant)(retryNum)
                executor.run(job.underlying, cronContext)
              }
              _ <- addNextExecutionToState(job, runInfo._1)
              _ <- logState()
              completed <- IO.fromFuture(IO(runInfo._2))
            } yield completed).recoverWith {
              case e: Throwable =>
                if(retryNum < maxRetry) {
                  val nextRetry = retryNum + 1
                  logger.debug(s"Job ${retryJob.underlying.id} has failed it's going to be retried. Retry number: $nextRetry")
                  runAndRetry(nextRetry)
                }
                else {
                  logger.debug(s"log stop retrying...")
                  IO.raiseError(e)
                }
            }

          logger.debug(s"Run job ${newJob.underlying.id} at ${scheduledAt.instant} with a delay of ${scheduledAt.delay}")
          for {
             _ <- addNextEventToState(job, scheduledAt.instant)
             _ <- logState()
             _ <- Timeout.applyF(scheduledAt.delay)
             _ <- runAndRetry(0)
             _ <- removeJobFromState(job)
          } yield ()
      }
  } yield ()
```

I'm not sure of this code though, it's just to give you an idea.

Reply:
done